lib/gemini.c

```diff
@@ -0,0 +1,296 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) Daniel Stenberg, <daniel@haxx.se>, et al.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ * SPDX-License-Identifier: curl
+ *
+ ***************************************************************************/
+
+#include "curl_setup.h"
+
+#if defined(USE_SSL) && !defined(CURL_DISABLE_GEMINI)
+
+#include <string.h>
+#include "urldata.h"
+#include "transfer.h"
+#include "connect.h"
+#include "cfilters.h"
+#include "sendf.h"
+#include "multiif.h"
+#include "select.h"
+#include "strdup.h"
+#include "url.h"
+#include "curl_printf.h"
+#include "gemini.h"
+#include "memdebug.h"
+
+static CURLcode gemini_setup_connection(struct Curl_easy *data,
+                                        struct connectdata *conn)
+{
+  (void)conn;
+  DEBUGASSERT(data->req.p.gemini == NULL);
+
+  data->req.p.gemini = calloc(1, sizeof(struct GEMINI));
+
+  if(!data->req.p.gemini)
+    return CURLE_OUT_OF_MEMORY;
+
+  return CURLE_OK;
+}
+
+static CURLcode gemini_connect(struct Curl_easy *data, bool *done)
+{
+  char *url = data->state.url;
+  struct GEMINI *gemini = data->req.p.gemini;
+
+  /* url is guaranteed to be non-NULL and non-empty */
+  DEBUGASSERT(url);
+
+  /* gemini is guaranteed to be non-NULL */
+  DEBUGASSERT(gemini);
+
+  /* url must not start with 0xFEFF */
+  if((unsigned char)url[0] == 0xFF && (unsigned char)url[1] == 0xFE) {
+    failf(data, "URL starts with 0xFEFF");
+    return CURLE_URL_MALFORMAT;
+  }
+
+  /* url must be 1024 bytes or less */
+  if(strlen(url) > 1024) {
+    failf(data, "URL is too long");
```

Comment:
How can gemini as a protocol accept URL as a general concept and yet refuse to work on URLs longer than 1024 bytes? Shouldn't the protocol rather reject too long file names or something? Where is this restriction specified?

Reply:
The specification directly calls out this restriction (https://gemini.circumlunar.space/docs/specification.html Section 2: Gemini Request). I'm moving the check to the URL parsing and calling the reference to the spec.

Reply:
I have since read the spec and I found that limit specified, thanks. I do have some additional questions on the UTF-8 encoding of the URL, because the limit is specified for the UTF-8 version and it's not at all clear to me what that means.

If there needs to be some encoding done, then the URL parser cannot enforce the limit since such an encoding process might change the length.